File: backend/src/services/event-spawner.service.ts
import { NodeDefinition } from '../models/node';
import { EventRepository } from '../repo/event-repo';
import { EventDefinition } from '../models/event-definition';

export interface EventSpec {
  name: string;
  min: number;
  max: number | null;
}

export class EventSpawnerService {
  private eventRepo = new EventRepository();

  /** Return legacy-shaped specs derived from the static JSON definitions. */
  async getSpecs(): Promise<EventSpec[]> {
    const defs = await this.eventRepo.getAll();
    return defs.map(d => ({
      name: d.type,
      min:  d.spawnRules.min,
      max:  d.spawnRules.max,
    }));
  }

  async validateCount(eventType: string, count: number): Promise<{ valid: boolean; reason?: string }> {
    const def = await this.eventRepo.getByType(eventType);
    if (!def) return { valid: false, reason: 'unknown event type' };
    const { min, max } = def.spawnRules;
    if (max !== null && count > max) return { valid: false, reason: `exceeds max ${max}` };
    if (count < min) return { valid: false, reason: `below min ${min}` };
    return { valid: true };
  }

  /**
   * Assign events to nodes while enforcing spawn caps read from events.json.
   * Nodes whose event count exceeds max are demoted to 'battle';
   * under-represented types are promoted from 'battle' nodes.
   */
  async assignEvents(nodes: NodeDefinition[]): Promise<NodeDefinition[]> {
    const defs = await this.eventRepo.getAll();
    // Build a lookup: event type -> spawn rules
    const rulesMap = new Map<string, { min: number; max: number | null }>();
    for (const d of defs) {
      rulesMap.set(d.type, { min: d.spawnRules.min, max: d.spawnRules.max });
    }

    // Count existing events
    const counts: Record<string, number> = {};
    nodes.forEach(n => {
      const t = n.event?.type ?? 'battle';
      counts[t] = (counts[t] || 0) + 1;
    });

    // Ensure max caps: demote excess to 'battle'
    for (const [type, rules] of rulesMap.entries()) {
      if (rules.max === null) continue;
      const current = counts[type] || 0;
      if (current > rules.max) {
        let toRemove = current - rules.max;
        for (const node of nodes) {
          if (toRemove <= 0) break;
          if (node.event?.type === type) {
            node.event.type = 'battle';
            counts[type] -= 1;
            counts['battle'] = (counts['battle'] || 0) + 1;
            toRemove -= 1;
          }
        }
      }
    }

    // Ensure min caps: promote 'battle' nodes into under-represented types
    for (const [type, rules] of rulesMap.entries()) {
      const current = counts[type] || 0;
      if (current < rules.min) {
        let need = rules.min - current;
        for (const node of nodes) {
          if (need <= 0) break;
          if ((node.event?.type ?? 'battle') === 'battle' && node.id !== 'start' && node.id !== 'end') {
            node.event = { type } as any;
            counts[type] = (counts[type] || 0) + 1;
            counts['battle'] = Math.max((counts['battle'] || 0) - 1, 0);
            need -= 1;
          }
        }
      }
    }

    return nodes;
  }
}

File: backend-csharp/TheGameIsOn.API/Services/EventSpawnerService.cs
using TheGameIsOn.API.Models;
using TheGameIsOn.API.Repositories;

namespace TheGameIsOn.API.Services;

public record EventSpec
{
    public string Name { get; init; } = string.Empty;
    public int Min { get; init; }
    public int? Max { get; init; }
}

public interface IEventSpawnerService
{
    Task<List<EventSpec>> GetSpecsAsync();
    Task<(bool Valid, string? Reason)> ValidateCountAsync(string eventType, int count);
    Task<List<NodeDefinition>> AssignEventsAsync(List<NodeDefinition> nodes);
}

public class EventSpawnerService : IEventSpawnerService
{
    private readonly IEventRepository _eventRepo;

    public EventSpawnerService(IEventRepository eventRepo)
    {
        _eventRepo = eventRepo;
    }

    public async Task<List<EventSpec>> GetSpecsAsync()
    {
        var defs = await _eventRepo.GetAllAsync();
        return defs.Select(d => new EventSpec
        {
            Name = d.Type,
            Min = d.SpawnRules.Min,
            Max = d.SpawnRules.Max
        }).ToList();
    }

    public async Task<(bool Valid, string? Reason)> ValidateCountAsync(string eventType, int count)
    {
        var def = await _eventRepo.GetByTypeAsync(eventType);
        if (def is null) return (false, "unknown event type");
        var rules = def.SpawnRules;
        if (rules.Max.HasValue && count > rules.Max.Value) return (false, $"exceeds max {rules.Max}");
        if (count < rules.Min) return (false, $"below min {rules.Min}");
        return (true, null);
    }

    /// <summary>
    /// Assign events to nodes while enforcing spawn caps.
    /// Excess nodes are demoted to "battle"; under-represented types are promoted.
    /// </summary>
    public async Task<List<NodeDefinition>> AssignEventsAsync(List<NodeDefinition> nodes)
    {
        var defs = await _eventRepo.GetAllAsync();
        var rulesMap = new Dictionary<string, (int Min, int? Max)>();
        foreach (var d in defs)
            rulesMap[d.Type] = (d.SpawnRules.Min, d.SpawnRules.Max);

        // Count existing events
        var counts = new Dictionary<string, int>();
        foreach (var n in nodes)
        {
            var t = n.Event.Type;
            counts[t] = counts.GetValueOrDefault(t) + 1;
        }

        // Enforce max caps: demote excess to "battle"
        foreach (var (type, rules) in rulesMap)
        {
            if (!rules.Max.HasValue) continue;
            var current = counts.GetValueOrDefault(type);
            if (current <= rules.Max.Value) continue;

            var toRemove = current - rules.Max.Value;
            for (var i = 0; i < nodes.Count && toRemove > 0; i++)
            {
                if (nodes[i].Event.Type == type)
                {
                    nodes[i] = nodes[i] with
                    {
                        Event = nodes[i].Event with { Type = "battle" }
                    };
                    counts[type] = counts.GetValueOrDefault(type) - 1;
                    counts["battle"] = counts.GetValueOrDefault("battle") + 1;
                    toRemove--;
                }
            }
        }

        // Enforce min caps: promote "battle" nodes
        foreach (var (type, rules) in rulesMap)
        {
            var current = counts.GetValueOrDefault(type);
            if (current >= rules.Min) continue;

            var need = rules.Min - current;
            for (var i = 0; i < nodes.Count && need > 0; i++)
            {
                if (nodes[i].Event.Type == "battle" && nodes[i].Id != "start" && nodes[i].Id != "end")
                {
                    nodes[i] = nodes[i] with
                    {
                        Event = nodes[i].Event with { Type = type }
                    };
                    counts[type] = counts.GetValueOrDefault(type) + 1;
                    counts["battle"] = Math.Max(counts.GetValueOrDefault("battle") - 1, 0);
                    need--;
                }
            }
        }

        return nodes;
    }
}
